src/Stack/PackageIndex.hs

```diff
@@ -247,12 +248,19 @@ updateIndexGit menv indexName' index gitUrl = do
                   $(mkRelDir "git-update")
                 acfDir = suDir </> repoName
             repoExists <- doesDirExist acfDir
-            unless repoExists
-                   (readProcessNull (Just suDir) menv "git" cloneArgs)
+            let doClone = readProcessNull (Just suDir) menv "git" cloneArgs
+            unless repoExists doClone
             isShallow <- doesFileExist $ acfDir </> $(mkRelDir ".git") </> $(mkRelFile "shallow")
             when isShallow $ do
               $logWarn "Shallow package index repo detected, transitioning to a full clone..."
-              readProcessNull (Just acfDir) menv "git" ["fetch", "--unshallow"]
+              let handleUnshallowError =
+                    C.handle $ \case
+                      ReadProcessException{} -> do
+                        $logInfo $ "Failed to convert to full clone, deleting and re-cloning."
+                        ignoringAbsence (removeDirRecur acfDir)
+                        doClone
+                      err -> throwM err
```

Comment:
Oh ok, I simply forgot to consider the possibility that `ReadProcessException` could be only one of several constructors for the same type.

I guess a constructor with the same name as the the type does suggest that it is the only constructor but I'm not aware of any preexisting convention. But yeah, either the constructor name or the type name should be changed.